## 423 - HW1
### Blessy Chinthapalli

## Exercise 1: Game Fun: Customer Acquisition through Digital Advertising

In [2]:
import pandas as pd

# Load the data
data = pd.read_excel('GameFun.xlsx')

data.head()

id  test  purchase   site  impressions  income  gender  gamer
0   1956     0         0  site1            0     100       1      0
1  45821     1         0  site1           20      70       1      0
2  59690     1         0  site1           22     100       1      0
3  18851     0         0  site1           13      90       1      0
4  60647     1         0  site1           12      60       1      0

In [3]:
from scipy import stats

# Compute means for test and control groups
group_means = data.groupby('test')[['income', 'gender', 'gamer']].mean()

# Compute the percentage difference in means
percent_diff = (group_means.loc[1] - group_means.loc[0]) / group_means.loc[0] * 100

# T-tests for income, gender, and gamer status
t_income = stats.ttest_ind(data[data['test'] == 1]['income'], data[data['test'] == 0]['income'], equal_var=False)
t_gender = stats.ttest_ind(data[data['test'] == 1]['gender'], data[data['test'] == 0]['gender'], equal_var=False)
t_gamer = stats.ttest_ind(data[data['test'] == 1]['gamer'], data[data['test'] == 0]['gamer'], equal_var=False)

print("Group_means :\n", group_means)
print("Percent_diff",percent_diff)
print(t_income)
print(t_gender)
print(t_gamer)


Group_means :
          income    gender     gamer
test                               
0     55.166012  0.647905  0.601823
1     54.938236  0.647289  0.601331
Percent_diff income   -0.412890
gender   -0.095049
gamer    -0.081720
dtype: float64
TtestResult(statistic=-1.5238055388794196, pvalue=0.12757124641315226, df=22666.753090524868)
TtestResult(statistic=-0.11806188043361146, pvalue=0.9060196778523593, df=22568.271654365355)
TtestResult(statistic=-0.09200032581416154, pvalue=0.9266985974352482, df=22564.381288798413)


Part 2

In [22]:
# Comaprision 1: Average purchase rate all customers

import pandas as pd

# Subset the data for the test and control groups
test_data = data[data['test'] == 1]
control_data = data[data['test'] == 0]

#  average purchase rate 
test_purchase_rate_all = test_data['purchase'].mean()
control_purchase_rate_all = control_data['purchase'].mean()

# absolute difference between the test and control purchase rates
diff_all = abs(test_purchase_rate_all - control_purchase_rate_all)

# Print the results
print("Comaprision 1: Average purchase rate all customers")
print("Average purchase rate for the test group (all customers):", round(test_purchase_rate_all, 4))
print("Average purchase rate for the control group (all customers):", round(control_purchase_rate_all, 4))
print("Absolute difference between the test and control purchase rates (all customers):", round(diff_all, 4))


Comaprision 1: Average purchase rate all customers
Average purchase rate for the test group (all customers): 0.0768
Average purchase rate for the control group (all customers): 0.0362
Absolute difference between the test and control purchase rates (all customers): 0.0406


In [24]:
# Comparison 2: Male vs Female customers 
print("Comparison 2: Male vs Female customers ")

import pandas as pd

## Male
# Subset the data for the test and control groups
test_data_male = test_data[test_data['gender'] == 1]
control_data_male = control_data[control_data['gender'] == 1]


test_purchase_rate_male = test_data_male['purchase'].mean()
control_purchase_rate_male = control_data_male['purchase'].mean()


diff_male = abs(test_purchase_rate_male - control_purchase_rate_male)

# Print the results
print("\nAverage purchase rate for the test group (male customers):", round(test_purchase_rate_male, 4))
print("Average purchase rate for the control group (male customers):", round(control_purchase_rate_male, 4))
print("Absolute difference between the test and control purchase rates (male customers):", round(diff_male, 4))

## Female
# Subset the data for the test and control groups
test_data_female = test_data[test_data['gender'] == 0]
control_data_female = control_data[control_data['gender'] == 0]


test_purchase_rate_female = test_data_female['purchase'].mean()
control_purchase_rate_female = control_data_female['purchase'].mean()


diff_female = abs(test_purchase_rate_female - control_purchase_rate_female)

# Print the results
print("\nAverage purchase rate for the test group (female customers):", round(test_purchase_rate_female, 4))
print("Average purchase rate for the control group (female customers):", round(control_purchase_rate_female, 4))
print("Absolute difference between the test and control purchase rates (female customers):", round(diff_female, 4))


Comparison 2: Male vs Female customers 

Average purchase rate for the test group (male customers): 0.0746
Average purchase rate for the control group (male customers): 0.0372
Absolute difference between the test and control purchase rates (male customers): 0.0374

Average purchase rate for the test group (female customers): 0.0809
Average purchase rate for the control group (female customers): 0.0344
Absolute difference between the test and control purchase rates (female customers): 0.0465


In [26]:
# Comparison 3: Gamers vs Non-Gamers Customers 
print("Comparison 3: Gamers vs Non-Gamers Customers ")
import pandas as pd


## Gamer
# Subset the data for the test and control groups
test_data_gamer = test_data[test_data['gamer'] == 1]
control_data_gamer = control_data[control_data['gamer'] == 1]

test_purchase_rate_gamer = test_data_gamer['purchase'].mean()
control_purchase_rate_gamer = control_data_gamer['purchase'].mean()

diff_gamer = abs(test_purchase_rate_gamer - control_purchase_rate_gamer)

# Print the results
print("\nAverage purchase rate for the test group (gamer customers):", round(test_purchase_rate_gamer, 4))
print("Average purchase rate for the control group (gamer customers):", round(control_purchase_rate_gamer, 4))
print("Absolute difference between the test and control purchase rates (gamer customers):", round(diff_gamer, 4))


## Non-gamer
# Subset the data for the test and control groups
test_data_non_gamer = test_data[test_data['gamer'] == 0]
control_data_non_gamer = control_data[control_data['gamer'] == 0]

test_purchase_rate_non_gamer = test_data_non_gamer['purchase'].mean()
control_purchase_rate_non_gamer = control_data_non_gamer['purchase'].mean()

diff_non_gamer = abs(test_purchase_rate_non_gamer - control_purchase_rate_non_gamer)

# Print the results
print("\nAverage purchase rate for the test group (non-gamer customers):", round(test_purchase_rate_non_gamer, 4))
print("Average purchase rate for the control group (non-gamer customers):", round(control_purchase_rate_non_gamer, 4))
print("Absolute difference between the test and control purchase rates (non-gamer customers):", round(diff_non_gamer, 4))


Comparison 3: Gamers vs Non-Gamers Customers 

Average purchase rate for the test group (gamer customers): 0.1045
Average purchase rate for the control group (gamer customers): 0.0354
Absolute difference between the test and control purchase rates (gamer customers): 0.0691

Average purchase rate for the test group (non-gamer customers): 0.0351
Average purchase rate for the control group (non-gamer customers): 0.0374
Absolute difference between the test and control purchase rates (non-gamer customers): 0.0023


In [28]:
#Comparison 4: Female Gamers vs Male Gamers
print("Comparison 4: Female Gamers vs Male Gamers")
import pandas as pd


## Female Gamers
test_data_female_gamer = test_data[(test_data['gamer'] == 1) & (test_data['gender'] == 0)]
control_data_female_gamer = control_data[(control_data['gamer'] == 1) & (control_data['gender'] == 0)]

test_purchase_rate_female_gamer = test_data_female_gamer['purchase'].mean()
control_purchase_rate_female_gamer = control_data_female_gamer['purchase'].mean()

diff_female_gamer = abs(test_purchase_rate_female_gamer - control_purchase_rate_female_gamer)

# Print the results
print("\nAverage purchase rate for the test group (female gamer customers):", round(test_purchase_rate_female_gamer, 4))
print("Average purchase rate for the control group (female gamer customers):", round(control_purchase_rate_female_gamer, 4))
print("Absolute difference between the test and control purchase rates (female gamer customers):", round(diff_female_gamer, 4))

## Male Gamers

test_data_male_gamer = test_data[(test_data['gamer'] == 1) & (test_data['gender'] == 1)]
control_data_male_gamer = control_data[(control_data['gamer'] == 1) & (control_data['gender'] == 1)]

test_purchase_rate_male_gamer = test_data_male_gamer['purchase'].mean()
control_purchase_rate_male_gamer = control_data_male_gamer['purchase'].mean()

diff_male_gamer = abs(test_purchase_rate_male_gamer - control_purchase_rate_male_gamer)

# Print the results
print("\nAverage purchase rate for the test group (male gamer customers):", round(test_purchase_rate_male_gamer, 4))
print("Average purchase rate for the control group (male gamer customers):", round(control_purchase_rate_male_gamer, 4))
print("Absolute difference between the test and control purchase rates (male gamer customers):", round(diff_male_gamer, 4))


Comparison 4: Female Gamers vs Male Gamers

Average purchase rate for the test group (female gamer customers): 0.1101
Average purchase rate for the control group (female gamer customers): 0.032
Absolute difference between the test and control purchase rates (female gamer customers): 0.0781

Average purchase rate for the test group (male gamer customers): 0.1014
Average purchase rate for the control group (male gamer customers): 0.0373
Absolute difference between the test and control purchase rates (male gamer customers): 0.0641


Expected revenue in the test vs. control for the following comparisons

In [30]:
import pandas as pd

# Constants
REVENUE_PER_CUSTOMER = 37.5
CREDIT_OFFERED = 25
NET_REVENUE_PER_CUSTOMER = REVENUE_PER_CUSTOMER - CREDIT_OFFERED  # $12.5

## Calculation for All Customers
# Number of purchases (assuming 'purchase' column is binary [0,1] where 1 indicates a purchase)
num_purchases_test_all = test_data['purchase'].sum()
num_purchases_control_all = control_data['purchase'].sum()

# Total revenue calculation
total_revenue_test_all = num_purchases_test_all * NET_REVENUE_PER_CUSTOMER
total_revenue_control_all = num_purchases_control_all * NET_REVENUE_PER_CUSTOMER

# Print results for all customers
print("All Customers")
print("Expected total net revenue for test group (all customers): $", total_revenue_test_all)
print("Expected total net revenue for control group (all customers): $", total_revenue_control_all)
print("Net revenue difference (test - control, all customers): $", total_revenue_test_all - total_revenue_control_all)

## Calculation for Female and Male Gamers
# Female Gamers
num_purchases_test_female_gamer = test_data_female_gamer['purchase'].sum()
num_purchases_control_female_gamer = control_data_female_gamer['purchase'].sum()

# Male Gamers
num_purchases_test_male_gamer = test_data_male_gamer['purchase'].sum()
num_purchases_control_male_gamer = control_data_male_gamer['purchase'].sum()

# Total revenue calculation - Female Gamers
total_revenue_test_female_gamer = num_purchases_test_female_gamer * NET_REVENUE_PER_CUSTOMER
total_revenue_control_female_gamer = num_purchases_control_female_gamer * NET_REVENUE_PER_CUSTOMER

# Total revenue calculation - Male Gamers
total_revenue_test_male_gamer = num_purchases_test_male_gamer * NET_REVENUE_PER_CUSTOMER
total_revenue_control_male_gamer = num_purchases_control_male_gamer * NET_REVENUE_PER_CUSTOMER

# Print results for female and male gamers
print("\nFemale and Male Gamers")
print("Expected total net revenue for test group (female gamers): $", total_revenue_test_female_gamer)
print("Expected total net revenue for control group (female gamers): $", total_revenue_control_female_gamer)
print("Net revenue difference (test - control, female gamers): $", total_revenue_test_female_gamer - total_revenue_control_female_gamer)

print("\nExpected total net revenue for test group (male gamers): $", total_revenue_test_male_gamer)
print("Expected total net revenue for control group (male gamers): $", total_revenue_control_male_gamer)
print("Net revenue difference (test - control, male gamers): $", total_revenue_test_male_gamer - total_revenue_control_male_gamer)


All Customers
Expected total net revenue for test group (all customers): $ 26975.0
Expected total net revenue for control group (all customers): $ 5412.5
Net revenue difference (test - control, all customers): $ 21562.5

Female and Male Gamers
Expected total net revenue for test group (female gamers): $ 8250.0
Expected total net revenue for control group (female gamers): $ 1012.5
Net revenue difference (test - control, female gamers): $ 7237.5

Expected total net revenue for test group (male gamers): $ 13812.5
Expected total net revenue for control group (male gamers): $ 2175.0
Net revenue difference (test - control, male gamers): $ 11637.5


Based on the promotional campaign analysis for Game-Fun, the financial outcomes indicate substantial gains in net revenue across different customer segments. The test groups, which were offered a $25 signing-up bonus to promote gaming subscription packages, demonstrated significantly higher revenue generation compared to the control groups. 

- **All Customers:** The test group generated $26,975 in net revenue, compared to $5,412.50 in the control group, resulting in a net revenue difference of $21,562.50. This suggests a robust overall effectiveness of the campaign in attracting new subscribers across the entire customer base.

- **Female Gamers:** Among this targeted subgroup, the test group's net revenue was $8,250, compared to $1,012.50 in the control group, with a revenue difference of $7,237.50. Female gamers responded exceptionally well to the promotion, indicating high potential for targeted marketing strategies.

- **Male Gamers:** The test group for male gamers brought in $13,812.50 in net revenue, while their control counterparts generated $2,175, showing a difference of $11,637.50. While the impact is also positive, it is slightly less pronounced compared to female gamers.

Given these results, it is advisable for Game-Fun to not only continue this promotional strategy but also consider optimizing it by focusing more on targeted segments, particularly female gamers. 

This group has shown the highest relative increase in revenue contribution, highlighting their greater susceptibility to such marketing incentives. Implementing targeted promotions could further enhance efficiency, maximizing the return on marketing expenditures while fostering higher engagement and loyalty within this segment. Continuous monitoring and tweaking of the campaign based on detailed performance analytics would ensure sustained growth and profitability.

## Exercise 2: Non-Compliance in Randomized Experiments

In [11]:
import pandas as pd

sommer_deger = pd.read_csv("sommer_deger.csv")

print(sommer_deger.head())


   instrument  treatment  outcome
0           0          0        0
1           0          0        0
2           0          0        0
3           0          0        0
4           0          0        0


In [13]:
print(sommer_deger)

       instrument  treatment  outcome
0               0          0        0
1               0          0        0
2               0          0        0
3               0          0        0
4               0          0        0
...           ...        ...      ...
23677           1          1        1
23678           1          1        1
23679           1          1        1
23680           1          1        1
23681           1          1        1

[23682 rows x 3 columns]


Comparing Survival Rates Based on Offered Vitamin A Shots

In [14]:
#  percent of babies whose mothers were offered Vitamin A shots for their babies and died
offered_died = sommer_deger[sommer_deger['instrument'] == 1]['outcome'].mean()

# percent of babies whose mothers were not offered Vitamin A shots for their babies and died
not_offered_died = sommer_deger[sommer_deger['instrument'] == 0]['outcome'].mean()

# difference in mortality
diff_mortality = offered_died - not_offered_died

print(f"Percent of babies whose mothers were offered Vitamin A shots for their babies and died: {offered_died * 100:.2f}%")
print(f"Percent of babies whose mothers were not offered Vitamin A shots for their babies and died: {not_offered_died * 100:.2f}%")
print(f"Difference in mortality: {diff_mortality * 100:.2f}%")

# Assumptions
print("\n Assumptions: The validity of using the difference in mortality rates between the Vitamin A shot and control groups as an estimate of the causal impact of the treatment on survival depends on the assumption that there are no confounding variables. This means that the only difference between the two groups is whether the mothers were offered the shot or not. If there are other factors that affect both the treatment assignment and the outcome variables, such as mother's education, socio-economic status, or past health history, the estimate may be biased and the method may not be appropriate.")

# Insights
print("\n Insights: The difference in mortality between babies whose mothers were offered Vitamin A shots and those who were not offered was -0.258%, indicating that the babies whose mothers were offered Vitamin A shots faced a lower mortality rate than those who were not offered. To ensure that this difference is a valid estimate of the causal impact of receiving Vitamin A shots on survival, we need to assume that the groups are comparable, all mothers who were offered Vitamin A shots actually took them, there were no spillover effects, SUTVA held, and there was no missing data.")


Percent of babies whose mothers were offered Vitamin A shots for their babies and died: 0.38%
Percent of babies whose mothers were not offered Vitamin A shots for their babies and died: 0.64%
Difference in mortality: -0.26%

 Assumptions: The validity of using the difference in mortality rates between the Vitamin A shot and control groups as an estimate of the causal impact of the treatment on survival depends on the assumption that there are no confounding variables. This means that the only difference between the two groups is whether the mothers were offered the shot or not. If there are other factors that affect both the treatment assignment and the outcome variables, such as mother's education, socio-economic status, or past health history, the estimate may be biased and the method may not be appropriate.

 Insights: The difference in mortality between babies whose mothers were offered Vitamin A shots and those who were not offered was -0.258%, indicating that the babies whose mot

Comparing Survival Rates Based on Actual Vitamin A Shot Receipt

In [17]:
#  percent of babies who received Vitamin A shots and died
received_died = sommer_deger[sommer_deger['treatment'] == 1]['outcome'].mean()

#  percent of babies who did not receive Vitamin A shots and died
not_received_died = sommer_deger[sommer_deger['treatment'] == 0]['outcome'].mean()

# Calculate the difference in mortality
diff_mortality = received_died - not_received_died

print(f"Percent of babies who received Vitamin A shots and died: {received_died * 100:.2f}%")
print(f"Percent of babies who did not receive Vitamin A shots and died: {not_received_died * 100:.2f}%")
print(f"Difference in mortality: {diff_mortality * 100:.2f}%")

# Assumptions
print("\nAssumptions: As discussed earlier, a valid estimate of the causal impact of Vitamin A shots on survival can be obtained under the assumption that there are no confounding factors. However, in the context of this particular dataset, there is a one-sided compliance issue, where some mothers in the treatment group did not receive the Vitamin A shot as intended. This non-compliance may lead to biased estimates and therefore, the difference in mortality rates between the treatment and control groups may not be the appropriate approach for this analysis. Alternative methods, such as instrumental variable analysis, may be required to account for the compliance issue and provide more reliable estimates of the causal effect of Vitamin A shots on survival.")

# Insights
print("\nInsights: The difference in mortality between those who received Vitamin A shots and those who did not is -0.644%, suggesting that babies who received Vitamin A shots experienced a lower mortality rate than those who did not receive the shot. To consider this estimate valid, we must assume randomization of subjects, no spillover effects, SUTVA, and no missing data.")


Percent of babies who received Vitamin A shots and died: 0.12%
Percent of babies who did not receive Vitamin A shots and died: 0.77%
Difference in mortality: -0.65%

Assumptions: As discussed earlier, a valid estimate of the causal impact of Vitamin A shots on survival can be obtained under the assumption that there are no confounding factors. However, in the context of this particular dataset, there is a one-sided compliance issue, where some mothers in the treatment group did not receive the Vitamin A shot as intended. This non-compliance may lead to biased estimates and therefore, the difference in mortality rates between the treatment and control groups may not be the appropriate approach for this analysis. Alternative methods, such as instrumental variable analysis, may be required to account for the compliance issue and provide more reliable estimates of the causal effect of Vitamin A shots on survival.

Insights: The difference in mortality between those who received Vitamin A s

Assessing Mortality Differences in Offered Vitamin A Shot

In [19]:
# only babies whose mothers were offered Vitamin A shots,

# Subset data for those offered Vitamin A shots
offered_vitamin = sommer_deger[sommer_deger['instrument'] == 1]

# Calculate mortality percentages for those who received and didn't receive the shot
vitaminA_shot_died = offered_vitamin[offered_vitamin['treatment'] == 1]['outcome'].mean() * 100
no_vitaminA_shot_died = offered_vitamin[offered_vitamin['treatment'] == 0]['outcome'].mean() * 100

# Difference in mortality
diff_mortality_offered = vitaminA_shot_died - no_vitaminA_shot_died
print(f"Difference in mortality: {diff_mortality_offered:.2f}%")
print(f"The difference in mortality between those who received Vitamin A shots and those who did not receive them is {diff_mortality_offered:.2f}%")

# Assumptions
print("\n Assumptions:  The validity of the estimate of the causal impact of Vitamin A shots on survival depends on the assumption that the only difference between the treatment and control groups is the receipt of the shot...")

# Insights
print("\n Insights: Comparing the mortality rates of babies who received Vitamin A shots versus those who did not, under the assumption that only mothers who were offered Vitamin A shots were considered, the difference in mortality is -1.28%...")


Difference in mortality: -1.28%
The difference in mortality between those who received Vitamin A shots and those who did not receive them is -1.28%

 Assumptions:  The validity of the estimate of the causal impact of Vitamin A shots on survival depends on the assumption that the only difference between the treatment and control groups is the receipt of the shot...

 Insights: Comparing the mortality rates of babies who received Vitamin A shots versus those who did not, under the assumption that only mothers who were offered Vitamin A shots were considered, the difference in mortality is -1.28%...


Wald Estimator

In [21]:
from statsmodels.formula.api import ols
import statsmodels.api as sm

# Calculate percentages directly using mean on the filtered data
percentage_babies_offered_died = offered_vitamin['outcome'].mean() * 100
percentage_babies_not_offered_died = sommer_deger[sommer_deger['instrument'] == 0]['outcome'].mean() * 100

# Offered and received Vitamin A
offered_received_vitamin = offered_vitamin['treatment'].mean() * 100

# Wald Estimate
wald_estimate = (percentage_babies_offered_died - percentage_babies_not_offered_died) / offered_received_vitamin
print(f"the wald estimate is {wald_estimate:.6f} percent")

# 4B
print("\nAssumptions: The Wald estimator can provide a valid estimate of the causal impact of vitamin A shots on survival under the assumption of no confounding and that the treatment assignment is independent of potential outcomes, given the covariates used in the analysis. However, if any of these assumptions are violated, such as the presence of unmeasured variables associated with both the treatment and outcome, the estimate may be biased and unreliable. Therefore, caution must be exercised, and alternative methods like IV's must be considered")

# Wald Standard Errors
model_intent_to_treat = ols('outcome ~ instrument', data=sommer_deger).fit(cov_type='HC0')
model_se = model_intent_to_treat.bse['instrument']
print(f"\nThe standard error for Intent-To-Treat estimate is {model_se}")

model_wald_est = ols('outcome ~ treatment', data=sommer_deger).fit(cov_type='HC0')
wald_est_se = model_wald_est.bse['treatment']
print(f"The standard error for Wald estimate is {wald_est_se}")

# Insights
print("The Wald SE is typically higher than the ITT error due to potential weak instrument problems or heterogeneous treatment effects across the population. The ITT estimate considers both compliers and non-compliers and may introduce additional noise leading to higher standard errors. On the other hand, the Wald estimate focuses on a more homogenous complier population and may result in lower standard errors. However, if the exclusion or independence assumptions are violated, the estimated standard errors might be biased and may not accurately capture the true variability in the treatment effect estimates. To address these issues, we may need additional information about potential confounding factors, implement randomized control trials or use alternative instruments or clustered standard errors estimates.")


the wald estimate is -0.003228 percent

Assumptions: The Wald estimator can provide a valid estimate of the causal impact of vitamin A shots on survival under the assumption of no confounding and that the treatment assignment is independent of potential outcomes, given the covariates used in the analysis. However, if any of these assumptions are violated, such as the presence of unmeasured variables associated with both the treatment and outcome, the estimate may be biased and unreliable. Therefore, caution must be exercised, and alternative methods like IV's must be considered

 The standard error for Intent-To-Treat estimate is 0.0009278269110806206
The standard error for Wald estimate is 0.0008211357141670643
The Wald standard error is typically higher than the ITT error due to potential weak instrument problems or heterogeneous treatment effects across the population...
